src/transaction/atomic_replication_helper.cpp

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include "atomic_replication_helper.hpp"
+
+#include "log_recovery_redo.hpp"
+#include "page_buffer.h"
+#include "system_parameter.h"
+
+namespace cublog
+{
+
+  /*********************************************************************
+   * atomic_replication_helper function definitions                    *
+   *********************************************************************/
+
+  template <typename T>
+  void atomic_replication_helper::add_atomic_replication_unit (THREAD_ENTRY *thread_p, TRANID tranid, log_lsa record_lsa,
+      LOG_RCVINDEX rcvindex, VPID vpid, log_rv_redo_context &redo_context, const log_rv_redo_rec_info<T> &record_info)
+  {
+#if !defined (NDEBUG)
+    if (!VPID_ISNULL (&vpid) && is_page_part_of_atomic_replication_sequence (vpid))
+      {
+	// page is already part of atomic replication
+	return;
```

Comment:
does this mean that; if this `vpid` is already part of an atomic replication unit, the current log record will not be processed?
IMHO, it should do the following:
- if the vpid is already part of an atomic replication
- the tranid of the existing atomic replication should be the same as the tranid of the new log record
If the new log record which is supposed to be part of an atomic replication has a different tranid; it means there are two transactions changing the same page which is a situation we should not see, but we need to make sure it does not occur